api/src/main/java/net/kyori/adventure/text/event/ClickCallback.java

```diff
@@ -0,0 +1,259 @@
+    return widen(original, type, null);
+  }
+
+  /**
+   * Perform an action for this event.
+   *
+   * @param audience the single-user audience who is attempting to execute this callback function.
+   * @since 4.13.0
+   */
+  void accept(final @NotNull T audience);
+
+  /**
+   * Filter audiences that receive this click callback.
+   *
+   * <p>Actions from audiences that do not match this predicate will be silently ignored.</p>
+   *
+   * @param filter the filter to test audiences with
+   * @return a filtered callback action
+   * @since 4.13.0
+   */
+  default @NotNull ClickCallback<T> filter(final @NotNull Predicate<T> filter) {
+    return this.filter(filter, null);
+  }
+
+  /**
+   * Filter audiences that receive this click callback.
+   *
+   * @param filter the filter to test audiences with
+   * @param otherwise the action to perform on the audience if the conditions are not met
+   * @return a filtered callback action
+   * @since 4.13.0
+   */
+  default @NotNull ClickCallback<T> filter(final @NotNull Predicate<T> filter, final @Nullable Consumer<? super Audience> otherwise) {
+    return audience -> {
+      if (filter.test(audience)) {
+        this.accept(audience);
+      } else if (otherwise != null) {
+        otherwise.accept(audience);
+      }
+    };
+  }
+
+  /**
+   * Require that audiences receiving this callback have a certain permission.
+   *
+   * <p>For audiences without permissions information, this test will always fail.</p>
+   *
+   * <p>Actions from audiences that do not match this predicate will be silently ignored.</p>
+   *
+   * @param permission the permission to check
+   * @return a modified callback
+   * @since 4.13.0
+   */
+  default @NotNull ClickCallback<T> requiringPermission(final @NotNull String permission) {
+    return this.requiringPermission(permission, null);
+  }
+
+  /**
+   * Require that audiences receiving this callback have a certain permission.
+   *
+   * <p>For audiences without permissions information, this test will always fail.</p>
+   *
+   * @param permission the permission to check
+   * @param otherwise the action to perform on the audience if the conditions are not met
+   * @return a modified callback
+   * @since 4.13.0
+   */
+  default @NotNull ClickCallback<T> requiringPermission(final @NotNull String permission, final @Nullable Consumer<? super Audience> otherwise) {
+    return this.filter(audience -> audience.getOrDefault(PermissionChecker.POINTER, ClickCallbackInternals.ALWAYS_FALSE).test(permission), otherwise);
+  }
+
+  /**
+   * Options to configure how a callback can be executed.
+   *
+   * @since 4.13.0
+   */
+  @ApiStatus.NonExtendable
+  interface Options extends Examinable {
+    /**
+     * Indicate that a callback should have unlimited uses.
+     *
+     * @since 4.13.0
+     */
+    int UNLIMITED_USES = -1;
```

Comment:
any reason why this one is in the options but the duration is in the callback? should really be both here

Reply:
thb probably best to put them both at top-level, just for better visibility and shorter field names